### PR TITLE
[GEP-18] Sync CA bundle to garden

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area security
/kind enhancement
/merge squash

**What this PR does / why we need it**:

- sync CA bundle to garden (new secret `<shoot-name>.ca-cluster`)
- sync CA bundle to kubeconfig secret (regressed by https://github.com/gardener/gardener/pull/5524)

This is done to decouple external components from the `ShootState` for retrieving the CA certificate.
Also, once CAs are rotated, there won't be a single CA cert in the `ShootState` anymore but multiple ones. The new dedicated `ca-cluster` secret will always contain the current CA bundle and is safe and easier to rely on in that sense.
(cc @holgerkoser @petersutter)

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/3292#issuecomment-1036058890

**Special notes for your reviewer**:

**Release note**:

```feature user
There is a new Secret for each Shoot in the corresponding Project Namespace (`<shoot-name>.ca-cluster`) which contains the current CA bundle for establishing trust to the Shoot's API server (see [documentation](https://github.com/gardener/gardener/blob/master/docs/usage/secrets_rotation.md#cluster-certificate-authority-bundle)).
```
```breaking operator
If you maintain `ResourceQuota` objects in the endusers' `Project` namespaces, make sure to increase the `secrets` quota, so that the new `<shoot-name>.ca-cluster` secret can be synced to the garden cluster (see [documentation](https://github.com/gardener/gardener/blob/master/docs/usage/secrets_rotation.md)).
```

### CODE REVIEW

File: docs/usage/secrets_rotation.md [hunk 1]
```diff
@@ -49,6 +49,18 @@ These client certificates are not revocable and thus a full credential rotation
 
 You can check the `.status.credentials.rotation.kubeconfig` field in the `Shoot` to see when the rotation was last initiated or last completed.
 
+#### Cluster Certificate Authority
```
Comment (rfranzke) [inline docs/usage/secrets_rotation.md]:
```suggestion
#### Cluster Certificate Authority Bundle
```

File: docs/usage/secrets_rotation.md [hunk 2]
```diff
@@ -49,6 +49,18 @@ These client certificates are not revocable and thus a full credential rotation
...
+
+*Name*: `<shoot-name>.ca-cluster`
+
+*Description*: Certificate Authority (CA) public key of the Cluster (`Secret` key: `ca.crt`).
```
Comment (rfranzke) [inline docs/usage/secrets_rotation.md]:
```suggestion
*Description*: Certificate Authority (CA) bundle of the Cluster (`Secret` key: `ca.crt`).
```

File: docs/usage/secrets_rotation.md [hunk 3]
```diff
@@ -49,6 +49,18 @@ These client certificates are not revocable and thus a full credential rotation
...
+
+This CA is used for signing serving certificates of the Shoot's API server. Hence, the public key contained in this `Secret` can be used to verify the API server's identity when communicating with its public endpoint (e.g. as `certificate-authority-data` in a Kubeconfig).
```
Comment (rfranzke) [inline docs/usage/secrets_rotation.md]:
```suggestion
This bundle can contain one or multiple CAs which are used for signing serving certificates of the Shoot's API server. Hence, the certificates contained in this `Secret` can be used to verify the API server's identity when communicating with its public endpoint (e.g. as `certificate-authority-data` in a Kubeconfig).
```

File: docs/usage/secrets_rotation.md [hunk 4]
```diff
@@ -49,6 +49,18 @@ These client certificates are not revocable and thus a full credential rotation
...
+This is the same public key that is also contained in the Kubeconfig's `certificate-authority-data` field.
```
Comment (rfranzke) [inline docs/usage/secrets_rotation.md]:
```suggestion
This is the same certificate that is also contained in the Kubeconfig's `certificate-authority-data` field.
```

File: docs/usage/secrets_rotation.md [hunk 5]
```diff
@@ -49,6 +49,18 @@ These client certificates are not revocable and thus a full credential rotation
...
+
+*Rotation*: Not supported yet, but work is in progress. See [gardener/gardener#3292](https://github.com/gardener/gardener/issues/3292) and [GEP-18](https://github.com/gardener/gardener/blob/release-v1.42/docs/proposals/18-shoot-CA-rotation.md) for more details.
+During an ongoing CA rotation operation, this `Secret` contains a certificate bundle of both the old and the new CA certificate.
```
Comment (rfranzke) [inline docs/usage/secrets_rotation.md]:
```suggestion
During an ongoing CA rotation operation, this `Secret` contains a certificate bundle of both the old and the new CA certificate.
```

I would drop this here (implementation detail). Better end-users are always expecting a bundle (multiple certs).

File: pkg/operation/botanist/secrets.go [hunk 6]
```diff
@@ -164,7 +164,18 @@ func (b *Botanist) generateCertificateAuthorities(ctx context.Context) error {
...
+	if !found {
+		return fmt.Errorf("secret %q not found", v1beta1constants.SecretNameCACluster)
+	}
+
+	return b.syncShootCredentialToGarden(
+		ctx,
+		gutil.ShootProjectSecretSuffixCACluster,
+		map[string]string{v1beta1constants.GardenRole: v1beta1constants.GardenRoleCACluster},
+		nil,
+		map[string][]byte{secretutils.DataKeyCertificateCA: caBundleSecret.Data[secretutils.DataKeyCertificateBundle]},
```
Comment (rfranzke) [inline pkg/operation/botanist/secrets.go:177]: Nit: Should we also use `DataKeyCertificateBundle` as key for the secret in the garden?

File: pkg/operation/botanist/kubeapiserver.go [hunk 7]
```diff
@@ -449,12 +450,21 @@ func (b *Botanist) DeployKubeAPIServer(ctx context.Context) error {
...
 	}
 
+	// add CA bundle as ca.crt to kubeconfig secret for backwards-compatibility
+	caBundleSecret, found := b.SecretsManager.Get(v1beta1constants.SecretNameCACluster)
+	if !found {
+		return fmt.Errorf("secret %q not found", v1beta1constants.SecretNameCACluster)
+	}
+
+	kubeconfigSecretData := userKubeconfigSecret.DeepCopy().Data
+	kubeconfigSecretData[secretutils.DataKeyCertificateCA] = caBundleSecret.Data[secretutils.DataKeyCertificateBundle]
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go:460]: Same question here

File: pkg/operation/botanist/secrets.go (hunk 6 above)
Comment (timebertt) [reply pkg/operation/botanist/secrets.go:177]: I went for `ca.crt` for consistency with the keys used in kubernetes secrets for the same content, e.g. service account secrets, `kube-root-ca.crt`, etc. All of these are using `ca.crt` as the key.
I also think it's quite commonly expected that these contain CA bundles and components typically read a whole bundle from these keys/files if present. Hence, I don't think `bundle.crt` is needed as a key to make this more obvious.
WDYT?

File: pkg/operation/botanist/kubeapiserver.go (hunk 7 above)
Comment (timebertt) [reply pkg/operation/botanist/kubeapiserver.go:460]: I think we should definitely keep it here for backwards-compatibility. 
If we change the key here, then we should rather drop this value at all and tell folks to read from the dedicated secret instead. In this case, they will need to adapt to the breaking change anyway, if they are already reading from this secret.

**Comment by timebertt** [comment]
I found a problem with this PR. At first, I thought the e2e job flaked, as I didn't see any test output in https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/5612/pull-gardener-e2e-kind/1504868779409018880.
However, when reproducing it locally, shoots are failing to be created with:
```
  Warning  ReconcileError  19s (x9 over 4m29s)   gardenlet  Flow "Shoot cluster reconciliation" encountered task errors: [task "Initializing secrets management" failed: retry failed with context deadline exceeded, last error: admission webhook "seed-restriction.gardener.cloud" denied the request: object does not belong to seed "local"]
```
Something seems to be wrong with our ginkgo timeouts or error reporting. It would be great to enhance that, so that folks don't need to reproduce this kind of error locally.

**Comment by timebertt** [comment]
I have also seen the following error again. I observed this multiple times already on first creation of shoots:
```
task "Persisting etcd encryption configuration in ShootState" failed: Operation cannot be fulfilled on shootstates.core.gardener.cloud "default-6zh2r": the object has been modified; please apply your changes to the latest version and try again
```

That's because there is no retry for the flow task or the actual patch in https://github.com/gardener/gardener/blob/7ac4b04feec409f3e5a5208cd06af9a10c755337/pkg/operation/operation.go#L579-L581
However, the secret handling of the etcd encryption secret is going to change with https://github.com/gardener/gardener/pull/5616. So we probably don't need to do anything about it for now.

**Comment by timebertt** [comment]
As there is no default resourcequota, the only thing we can/should do for adapting to the new secret is to leave a release note of category "breaking" for operators.
After adapting the seed admission and auth, I marked the PR as ready accordingly.

**Review by rfranzke — APPROVED** [review]

**Pull request merged**